src/drivers/imx/dummy-dma.c

```diff
@@ -45,54 +134,302 @@ static int dummy_dma_status(struct dma_chan_data *channel,
+	int ret = 0;
+
+	spin_lock_irq(channel->dma->lock, flags);
+
+	if (!config->elem_array.count) {
+		trace_dummydma_error("dummy-dmac: %d channel %d no DMA descriptors",
+				     channel->dma->plat_data.id,
+				     channel->index);
+
+		ret = -EINVAL;
+		goto out;
+	}
+
+	channel->direction = config->direction;
+	/* TODO Should we keep this check? */
+	if (config->direction != DMA_DIR_HMEM_TO_LMEM &&
+	    config->direction != DMA_DIR_LMEM_TO_HMEM) {
+		/* Shouldn't even happen though */
+		trace_dummydma_error("dummy-dmac: %d channel %d invalid direction %d",
+				     channel->dma->plat_data.id, channel->index,
+				     config->direction);
+		ret = -EINVAL;
+		goto out;
+	}
+	channel->desc_count = config->elem_array.count;
+	ch->elems = &config->elem_array;
+	ch->sg_elem_curr_idx = 0;
+
+	channel->status = COMP_STATE_PREPARE;
+out:
+	spin_unlock_irq(channel->dma->lock, flags);
+	return ret;
 }
 
 /* restore DMA context after leaving D3 */
 static int dummy_dma_pm_context_restore(struct dma *dma)
 {
+	/* Virtual device, no hardware registers */
 	return 0;
 }
 
 /* store DMA context after leaving D3 */
 static int dummy_dma_pm_context_store(struct dma *dma)
 {
+	/* Virtual device, no hardware registers */
 	return 0;
 }
 
 static int dummy_dma_set_cb(struct dma_chan_data *channel, int type,
 		void (*cb)(void *data, uint32_t type, struct dma_cb_data *next),
 		void *data)
 {
+	channel->cb = cb;
+	channel->cb_data = data;
+	channel->cb_type = type;
 	return 0;
 }
 
+/* This function extracts the next element in the channel's element list
+ * (that list is set at configure time). It is used for continuing the
+ * copy flow.
+ */
+static bool get_next_elem(struct dma_chan_data *channel,
+			  struct dma_cb_data *next)
+{
+	struct dma_chan_pdata *ch = dma_chan_get_data(channel);
+	int curr_index = ch->sg_elem_curr_idx++;
+
+	if (curr_index >= channel->desc_count)
+		return false;
+
+	next->elem = ch->elems->elems[curr_index];
+	return true;
+}
+
+/* Perform a single elem copy, call the IRQ callback and prepare for the
+ * next copy
+ * Returns how many bytes were actually copied, either as a positive
+ * number to allow continuing or as a negative number to stop
+ * immediately (due to callback request).
+ */
+static int docopy(struct dma_chan_data *ch, struct dma_cb_data *next)
+{
+	struct dma_chan_pdata *pdata = dma_chan_get_data(ch);
+	/* Perform a single elem copy */
+	void *dest = (void *)next->elem.dest;
+	void *src = (void *)next->elem.src;
+	size_t size = next->elem.size;
+
+	/* Do the copy */
+	/* Note that we don't know the real buffer size,
+	 * we assume `size` is correct
+	 */
+	memcpy_s(dest, size, src, size);
+
+	/* For status, point to immediately after the last copy */
+	pdata->r_pos = next->elem.src + size;
+	pdata->w_pos = next->elem.dest + size;
+
+	if (!(ch->cb_type & DMA_CB_TYPE_IRQ))
+		return -size; /* Callback not applicable, end for now */
+
+	/* If the callback asks for nothing special just load the next
```

Comment:
We need to add a comment like this: syncrhonously call DMA callback to finish current sg element processing.

Reply:
Changed a few things around there, please re-review.